### DISCUSSION TITLE
How to overlay an image onto a background transparently?

### DISCUSSION THREAD

**Discussion by pvonmoradi** [ImageMagick/ImageMagick#4851] [Help] [answered]
I have a foreground image which parts of it is fully-transparent. I want to overlay this image onto a background in a way that:
1- In areas where the fg image was fully-transparent, the final image should only have the background.
2- In areas where the fg image was opaque, the final image should be the overlay of these parts over background in a transparent way (like 70%).
Example:
fg-opaque: `convert -size 256x256 xc:blue b256.png`
fg-transparent: `convert -size 256x256 xc:transparent t256.png`
fg-image:`montage b256.png  t256.png -mode concatenate -background none bt512.png` (left part: opaque, right part: fully-transparent)
bg-image:`convert -size 512x512 xc:green back_512.png` (fully opaque green background)

Here is what I tried:  
result1:`composite -compose dst_over back_512.png bt512.png out.png`
This method overlays the fg-image on bg-image completely. It meets (1) but does not meet (2).

If I use result2: `composite -compose dst_over -dissolve 70 back_512.png bt512.png out.png`, it would meet (2) but not (1).
I want the left part be  transparent (70% blue on green), but I want the right part to be green completely (since the right part of fg-image is fully-transparent, and if we overlay a fully-transparent thing onto an opaque thing, the result should be that background opaque thing logically)

result1:  
![image](https://user-images.githubusercontent.com/1058151/154754992-17c6013a-f6ae-4e47-a2d8-b155abcf1d18.png)

result2:  
![image](https://user-images.githubusercontent.com/1058151/154755087-a24f773f-f05d-4c7f-bb58-bf50fdf05c78.png)

**Comment by fmw42** [comment]
Please post your actual input images separately.  What is your IM version and Platform?  You should use convert .... -composite rather than composite.  It is easier to do other things in the command line. 

See https://legacy.imagemagick.org/Usage/compose/#compose

**Reply by pvonmoradi** [reply]
`Version: ImageMagick 6.9.7-4 Q16 x86_64 20170114` on Linux
Actual images are just like `fg-image` and `bg-image` :

fg: https://transfer.sh/2eJd2R/bt512.png

bg: https://transfer.sh/z3WNNc/back_512.png

**Comment by fmw42** [comment, marked as answer]
Does this do what you want?  It requires IM 7 and will not work in IM 6 unless you do the computation for the shave amount first, put it in a variable and pass the variable to the command line.

`magick back_512.png bt512.png -set option:shave "%[fx:0.5*(u.h-v.h)]" -gravity center -compose over -composite -shave "0x%[shave]" -bordercolor none -border "0x%[shave]" x.png`

![x](https://user-images.githubusercontent.com/13458218/154762470-4e109d89-deae-48a2-880c-daae0c0c2d19.png)

**Reply by pvonmoradi** [reply]
Actually `composite -compose over -dissolve 70  bt512.png back_512.png out.png` worked! My problem was using `dst-over` instead of simply using `over`. Your [duff-porter link](https://legacy.imagemagick.org/Usage/compose/#duff-porter) helped solving this. Thanks!
